lib/read.js

```diff
@@ -189,137 +310,51 @@ class ReadSql extends AdapterRead {
-                this.trigger('error', err);
-            }
-        })
-        .then(() => {
-            var points = this.pointsBuffer;
-            this.pointsBuffer = [];
-            return {
-                points: points,
-                readEnd: this.timeField && !this.debugOption ? to : new JuttleMoment(Infinity)
-            };
-        });
-    }
-
-    paginateBuckets(original_query) {
-        var self = this;
-        var timeField = self.timeField || 'time';
-        var buckets = self.batchBuckets;
-        return Promise.mapSeries(buckets.slice(1), function(bucket, i, len) {
-            var query = original_query.clone();
-            query = query.where(timeField, '>=', new Date(buckets[i].valueOf()));
-            query = query.where(timeField, '<', new Date(bucket.valueOf()));
-
-            var additionalFields = {};
-            bucket.epsilon = true;
-            additionalFields[timeField] = bucket;
-
-            return self.paginate(query, additionalFields);
-        });
-    }
-
-    //the raw query response format can be different based on client
-    handleRawResponse(points) {
-        return points;
-    }
-
-    paginate(query, additionalFields) {
-        this.logger.debug('executing paginated query');
+    executeQuery(query, to) {
+        this.logger.debug('executing query', query.toString());
 
         return query.then((points) => {
-            var nextQuery;
-
-            if (additionalFields) {
-                _.each(points, function(pt) {
-                    _.extend(pt, additionalFields);
-                });
-            }
             if (points.length < this.fetchSize) {
                 //no more pagination necessary
-                return this.formatAndBuffer(points);
+                return {
+                    points: points,
+                    readEnd: this.timeField ? to : new JuttleMoment(Infinity)
+                };
             }
-
             //perform time-based pagination if timeField is indicated
             if (this.timeField) {
-                nextQuery = this.processPaginatedResults(query, points, this.timeField);
-            } else {
-                this.formatAndBuffer(points);
-                this.offsetCount += points.length;
-                this.logger.debug('next pagination based on offset ', this.offsetCount);
-                nextQuery = query.offset(this.offsetCount).limit(this.getNextLimit());
+                var res = this.processPaginatedResults(points, this.timeField);
+                res.readEnd = new JuttleMoment({rawDate: new Date(res.borderValue)});
+                return res;
             }
 
-            if (nextQuery && this.getNextLimit() > 0) {
-                return this.paginate(nextQuery);
-            }
+            //offset pagination
+            this.offsetCount += points.length;
+            return {
+                points: points,
+                readEnd: null //read again
+            };
+        })
+        .then((res) => {
```

Comment:
Not if you do it on 322 or can you not call formatPoints before processPaginatedResults?